src/extension/baser.ts

```diff
@@ -0,0 +1,171 @@
+            await workspace.openTextDocument(Uri.parse(path));
+        } catch (error) {
+            return false;
+        }
+        return true;
+    }
+
+    private activeInfoMessages = new Set<string>() // Prevent repeat message animations when arrowing through many results with the same uri.
+    public async translateArtifactToLocal(artifactPath: string) { // Retval is validated.
+        // Temp.
+        if (artifactPath.startsWith('sarif:')) return artifactPath;
+        const validateUri = async () => {
+            // Cache
+            if (this.validatedPathsArtifactToLocal.has(artifactPath))
+                return this.validatedPathsArtifactToLocal.get(artifactPath);
+
+            // File System Exist
+            if (await this.pathExists(artifactPath))
+                return artifactPath;
+
+            // Known Bases
+            for (const [artifactBase, localBase] of this.basesArtifactToLocal) {
+                if (!artifactPath.startsWith(artifactBase)) continue; // Just let it fall through?
+
+                const normalizedArtifactPath = `${artifactPath.startsWith('/') ? '' : '/'}${artifactPath}`;
+                const localPath = normalizedArtifactPath.replace(artifactBase, localBase);
+
+                if (await this.pathExists(localPath)) {
+                    this.updateValidatedPaths(artifactPath, localPath);
+                    return localPath;
+                }
+            }
+
+            { // API-injected baseUris
+                const localPath = await this.tryUriBases(artifactPath);
+                if (localPath) return localPath;
+            }
+
+            // Distinct Project Items
+            const {file} = artifactPath;
+            if (this.distinctLocalNames.has(file) && this.store.distinctArtifactNames.has(file)) {
+                const localPath = this.distinctLocalNames.get(file)!; // Not undefined due to surrounding if.
+                this.updateValidatedPaths(artifactPath, localPath);
+                this.updateBases(artifactPath.split('/'), localPath.split('/'));
+                return localPath;
+            }
+
+            // Open Docs
+            for (const doc of workspace.textDocuments) {
+                const localPath = doc.uri.path;
+                if (localPath.file !== artifactPath.file) continue;
+                this.updateValidatedPaths(artifactPath, localPath);
+                this.updateBases(artifactPath.split('/'), localPath.split('/'));
+                return localPath;
+            }
+
+            return ''; // Can't find uri.
+        };
+
+        let validatedUri = await validateUri();
+        if (!validatedUri && !this.activeInfoMessages.has(artifactPath)) {
+            this.activeInfoMessages.add(artifactPath);
+            const choice = await window.showInformationMessage(`Unable to find '${artifactPath.split('/').pop()}'`, 'Locate...');
+            this.activeInfoMessages.delete(artifactPath);
+            if (choice === 'Locate...') {
```

Comment:
>choice === 'Locate...' [](start = 16, length = 22)

Let's either compare the return value of showInformationMessage to undefined "if (choice)" (since there is only one argument) or defined a message item object locally can compare the return to that.

You should never compare results to a user displayable string.

Reply:
Agreed. Fixed.

---
In reply to: [441638778](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441638778) [](ancestors = 441638778)